### PR TITLE
Give an option to not connect offline Discord users

### PR DESCRIPTION
**Pull request by Sazpaimon**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -41,6 +41,10 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
                 #webhooks_agent: false
                 webhooks_agent: Webhooks
                 #webhooks_agent: Webhooks!webhooks@discord/webhooks-agent
+                # If disabled, users that are marked as "Invisible" or "Offline" will not be joined to 
+                # linked channels until they come online. Useful if you have many offline users compared 
+                # to online ones. Note that if this is disabled, PMs cannot be sent to offline Discord users
```
Comment (jlu5) [inline README.md:46]:
```yaml
                    # This defaults to true if not set.
```

File: protocols/discord.py [hunk 2]
```diff
@@ -227,6 +227,10 @@ def _burst_new_client(self, guild, member, pylink_netobj):
             log.debug('(%s) Not bursting user %s as their data is not ready yet', self.protocol.name, member)
             return
 
+        if not pylink_netobj.serverdata.get('join_offline_users', True) and \
+                member.user.presence in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE, UNSET]:
```
Comment (jlu5) [inline protocols/discord.py]: Could we move this into a helper function like `_should_burst_user()`? I believe it's possible to invert the logic as follows:

```python
return pylink_netobj.serverdata.get('join_offline_users', True) or \
    member.user.presence not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE, UNSET]
```

**Review by jlu5 — CHANGES_REQUESTED** [review]
I'm a bit concerned about desyncs here:

1) If the `join_offline_users` option value changes while a guild is connected, things could go badly. For the sake of simplicity, we could perhaps store that value when we initialize a subserver and don't allow it to be changed without a reconnect.

2) I believe the check in `_update_channel_presence` should be updated to silently ignore missing users if this option is disabled, instead of raising an error.

File: README.md [hunk 3]
```diff
@@ -41,6 +41,12 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
...
+                # Changes to this setting take place on restart
```
Comment (jlu5) [inline README.md]: Mm, more correct would be:

```yaml
    # Changes to this setting require a reconnect to apply.
```

File: protocols/discord.py [hunk 4]
```diff
@@ -482,7 +495,14 @@ def _update_user_status(self, guild_id, uid, presence):
...
     @Plugin.listen('PresenceUpdate')
     def on_presence_update(self, event, *args, **kwargs):
-        self._update_user_status(event.guild_id, event.presence.user.id, event.presence)
+        self._update_user_status(event.guild, event.presence.user.id, event.presence)
+
+    @staticmethod
+    def _should_burst_user(pylink_netobj, user):
+        return not pylink_netobj.join_offline_users and \
+               user.presence != UNSET and \
+               user.presence.status not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE]
```
Comment (jlu5) [inline protocols/discord.py]: Could we rewrite this as an `or`? I think it's easier to read

```python
    # Burst users only if join_offline_users is enabled or their status is online/away/DND
    return pylink_netobj.join_offline_users or (user.presence != UNSET and \
        user.presence.status not in [DiscordStatus.OFFLINE, DiscordStatus.INVISIBLE])
```

File: README.md [hunk 5]
```diff
@@ -41,6 +41,12 @@ You can also install these dependencies via pip (for Python 3) using: `pip3 inst
...
+                # This defaults to true if not set.
```
Comment (jlu5) [inline README.md]: And we can put this with the line above

File: protocols/discord.py (hunk 4 above)
Comment (Sazpaimon) [reply protocols/discord.py]: I personally prefer if statements being only one level deep, but I can swing either way, here

Comment (jlu5) [reply protocols/discord.py]: I think logically clearer is better

**Pull request merged**